packages/@aws-cdk/aws-codebuild/lib/project.ts

```diff
@@ -901,8 +901,7 @@ export class Project extends ProjectBase {
       },
     }));
 
-    const policy = new iam.Policy(this, 'PolicyDocument', {
-      policyName: 'CodeBuildEC2Policy',
+    const policy = this.role.createAndAttachPolicy('CodeBuildVpcPolicy', {
```

Comment:
The fact that this can return nothing is a bit of a gotcha and feels like an unnatural API.

Can it not return a special kind of `Policy` object that doesn't render to CFN, or something?

Reply:
Maybe the change should be that unattached `Policy` objects just don't render? This new method feels unnatural.

Reply:
Why? Why is returning something optionally "unnatural"...?

Returning a magical `Policy` that won't render if it's unattached seems a lot more unnatural to me (we don't have a single resource in the library that behaves that way - doesn't render under some conditions), and also won't work in this case - we are making the `Project` a dependency of the `Policy`, so we can't not render it (the resulting template will be invalid).

Reply:
> Why? Why is returning something optionally "unnatural"...?

How is it to be expected that a function that is called `createThing` doesn't actually create and return a thing? What I don't like about this is:

* Consumers have to write additional ifs. I will fight against any design that requires my consumers to add an `if` to their code.
* We could only conceivably have typed this function this way once we have the requirement that there may be a class which *doesn't* do the thing. If we had made this function BEFORE we had the requirement of the no-op class, the API would have been locked in to the normal API (which always returns a thing) and then we wouldn't have been able to add this in this way. That leads me to think we are confusing layers of functionality.

> we don't have a single resource in the library that behaves that way - doesn't render under some conditions

We do have objects that render differently under certain conditions (references turn into `{ Ref: xxx }` or `{ Fn::ImportValue: Yyy }` under certain conditions) so that is not without precedent. And in fact, we also have policy objects that automatically get added to the document if they're non-empty (see `BucketPolicy`, `QueuePolicy`, `TopicPolicy`).

Automatic adding on non-empty is roughly what I'm asking for here as well. The difference being that in the existing case, `bucket.addToResourcePolicy()` adds the `AWS::IAM::Policy` resource to the template, and in this case it would be `policy.addToPolicy()` that adds the `AWS::IAM::Policy` to the template.

Yes, the *letter* of the code will be different (because the class structure is slightly different), but I would argue it's in the same *spirit*.

And in any case, CDK is intended to make it easier to work with CloudFormation templates, not push the same problems and need for case analysis down to the user if we can deal with them.

> we are making the `Project` depend on the `Policy`, so we can't not render it

Obviously we would have to not render the dependency in the case we're not rendering the resource. Yes, I understand it is more work, but it's not like that is impossible to do.

Reply:
> Consumers have to write additional ifs. I will fight against any design that requires my consumers to add an if to their code.

That's a noble point, only slightly undercut by the fact that we're adding a method to an interface with an existing method `addToPolicy(statement: PolicyStatement): boolean`...

> Obviously we would have to not render the dependency in the case we're not rendering the resource. Yes, I understand it is more work, but it's not like that is impossible to do.

How? `CfnResource.addDependsOn(resource)` adds `resource` to a `dependsOn` Set field of `CfnResource`. How would you not render that?

And other cases - what if somebody does `CfnOutput(this, 'Output', { value: policy.ref })`? How could we possibly not render that?

Reply:
So probably this is a bad idea for `CfnPolicy`, but it could be true for `Policy`.

As for the second question, we could render it as `{ Ref: AWS::NoValue }`.

As for `addDependsOn()`... that is an issue. We would really have to use `addDependency`, which has a mechanism (in `IDependable`) to control what gets depended on when you take a dependency on a thing.

Reply:
I just spent basically the entire day implementing `LazyPolicy` (which I believe is basically what you asked). And it was horrible. Feel free to judge yourself (I've updated the PR with what I came up with).

We turned a clean, simple API into a Frankenstein monster full of hacks, throwing exceptions, and weird edge cases. I don't even know how to write the docs for this new method, it's so convoluted what it does now. And I'm still not able to actually make it work - the build fails on the `codebuild` package. 

I really urge you to reconsider. I think we're going down the wrong path here.